aml: add v0.2, cuda,opencl,hip support

perarnau opened the pull request:
Add release v0.2 of the AML package, deprecate v0.1, and add support for OpenCL, HIP, and CUDA variants of the library. Also update repo and release URL, as the previous one is not accessible anymore.

AML also supports level_zero (intel), but there's no package for it in Spack yet.

Note: previous PR (#27763) stalled and was closed after the person in charge left the group. Some fixes were applied since.

perarnau committed:
aml: add v0.2, cuda,opencl,hip support

Add release v0.2 of the AML package, deprecate v0.1, and add support for
OpenCL, HIP, and CUDA variants of the library. Also update repo and
release URL, as the previous one is not accessible anymore.

Note: AML also supports level_zero (intel), but there's no package for
it in Spack yet.

alalazo commented on var/spack/repos/builtin/packages/aml/package.py:
```diff
@@ -9,60 +9,122 @@
 
 class Aml(AutotoolsPackage):
-    """AML: Building Blocks for Memory Management."""
+    """AML is a memory management library designed to ease the use of complex
```
Is there a reason not to inherit from `CudaPackage` or `RocmPackage`?

perarnau replied:
We don't depend on the compilers, we only need to link to the underlying libraries. I don't mind inheriting from both Cuda and ROCm but:
- the cuda_arch variant is not going to be used by the library
- it's unclear how I can use ROCmPackage to allow a hip variant running on top of the nvidia runtime
- the ROCmPackage is going to force dependencies that we don't need

alalazo replied:
Thanks, that's fair.

alalazo commented on var/spack/repos/builtin/packages/aml/package.py:
```diff
@@ -9,60 +9,122 @@
+            description="Support for memory operations on top of CUDA.")
+    variant('hwloc', default=False,
+            description="Enable feature related to topology management")
+    variant('openmp-targets', values=disjoint_sets(('spir64',)),
+            description="Intel OpenMP target architecture setting.")
+    variant('hip-platform', values=disjoint_sets(('amd', 'nvidia')),
+            description="HIP backend platform.")
```
Why are we using `disjoint_sets` here, and declaring only one set? If there is a single set, a multi-valued variant would fit better.[^1]

[^1]: To clarify the comment,`disjoint_set` allows to select multiple values from the same set, but forbids having values from different sets.

perarnau replied:
I removed it as a variant and just enabled it when the compiler is `%oneapi`.

alalazo reviewed (changes requested):
Apologies for the delay in replying. I have one minor request, but otherwise this LGTM.

perarnau committed:
aml: add oneapi-level-zero support

Also changes openmp flags to force compatibility when compiling with
intel-oneapi compilers.

perarnau committed:
Merge branch 'develop' into package/aml-0.2

alalazo commented:
@spackbot fix style

alalazo reviewed (dismissed):
Can you fix style?

perarnau committed:
aml: style fixes

perarnau committed:
aml: don't quote openmp extra flags in spack

alalazo commented on var/spack/repos/builtin/packages/aml/package.py:
```diff
@@ -9,64 +9,126 @@
+    variant("hwloc", default=False, description="Enable feature related to topology management")
+    variant(
+        "hip-platform",
+        values=disjoint_sets(("amd", "nvidia")),
```
This seems wrong, since there is a single set in the declaration. A `disjoint_set` is usually employed like:
https://github.com/spack/spack/blob/6ab6058ab4ff046115f71d501c8ef20b9bc05ae3/var/spack/repos/builtin/packages/mvapich2/package.py#L75-L83

to mean that you can get e.g. `process_managers=hydra,gforker` but not  `process_managers=hydra,slurm`, i.e. you cannot combine together elements of different sets.

perarnau replied:
you're right. How do you want me to proceed since this one was merged ? new PR ?

tgamblin replied:
new PR is good.  @eugeneswalker FYI
